audiobook_generator/tts_providers/azure_tts_provider.py

```diff
@@ -0,0 +1,181 @@
+            logger.debug(f"SSML: [{ssml}]")
+
+            for retry in range(MAX_RETRIES):
+                self.auto_renew_access_token()
+                headers = {
+                    "Authorization": f"Bearer {self.access_token}",
+                    "Content-Type": "application/ssml+xml",
+                    "X-Microsoft-OutputFormat": self.output_format,
+                    "User-Agent": "Python",
+                }
+                try:
+                    logger.info(
+                        "Sending request to Azure TTS, data length: " + str(len(ssml))
+                    )
+                    response = requests.post(
+                        self.TTS_URL, headers=headers, data=ssml.encode("utf-8")
+                    )
+                    response.raise_for_status()  # Will raise HTTPError for 4XX or 5XX status
+                    logger.info(
+                        "Got response from Azure TTS, response length: "
+                        + str(len(response.content))
+                    )
+                    audio_segments.append(io.BytesIO(response.content))
+                    break
+                except requests.exceptions.RequestException as e:
+                    logger.warning(
+                        f"Error while converting text to speech (attempt {retry + 1}): {e}"
+                    )
+                    if retry < MAX_RETRIES - 1:
+                        sleep(2 ** retry)
+                    else:
+                        raise e
+
+        with open(output_file, "wb", encoding='utf-8') as outfile:
+            for segment in audio_segments:
+                segment.seek(0)
+                outfile.write(segment.read())
+
+        set_audio_tags(output_file, audio_tags)
+
+    def get_break_string(self):
+        return " @BRK#"
+
+    def get_output_file_extension(self):
+        if self.output_format.startswith("amr"):
+            return "amr"
+        elif self.output_format.startswith("ogg"):
+            return "ogg"
+        elif self.output_format.endswith("truesilk"):
+            return "silk"
+        elif self.output_format.endswith("pcm"):
+            return "pcm"
+        elif self.output_format.startswith("raw"):
+            return "wav"
+        elif self.output_format.startswith("webm"):
+            return "webm"
+        elif self.output_format.endswith("opus"):
+            return "opus"
+        elif self.output_format.endswith("mp3"):
+            return "mp3"
+        else:
+            raise NotImplementedError(f"Unknown file extension for output format: {self.output_format}")
+
+    def validate_config(self):
+        # TODO: Need to dig into Azure properties, im not familiar with them, but look at OpenAI as ref example
+        pass
+
+    def estimate_cost(self, total_chars):
+        return math.ceil(total_chars / 1000000) * self.price
```

Comment:
Seems wrong. Should be `math.ceil(total_chars / 1000000  * self.price)`? Or the result could be complete different (much higher than the actual value).

Reply:
no its correct,
Math.ceil is round the number to higher number, for example:
if we round 1.1 - it will be 1
but if we ceil 1.1 - it will be 2

So what I'm doing here is:

1. divide all chars by 1000000 and round to next full number
2. and only then multiply by price

So you will get higher bound estimate, like max price, but the actual book will be a bit cheaper.

Reply:
Assuming there is a book with 534,800 characters, according to your code, math.ceil(534800 / 1000000) * 16 = 16, indicating an estimated cost of 16. In reality, it would only cost less than 9 dollars. This difference is too significant. I understand that you calculated based on "$16 per 1M characters", but I don't believe Microsoft would charge 32 dollars for just 1.01 million characters.

Reply:
hmmm ye, you're right, for openAI estimate is way more accurate, ill improve for azure!

Reply:
fixed